core/src/main/java/org/infinispan/distribution/TriangleOrderManager.java

```diff
@@ -0,0 +1,187 @@
+      return senderSegmentMap.computeIfAbsent(segmentId, integer -> new SenderSequencer(topologyId)).next(topologyId);
+   }
+
+   private boolean checkIfNext(int segmentId, int topologyId, long sequenceNumber) {
+      return receiverSegmentMap.computeIfAbsent(segmentId, integer -> new ReceiverSequencer(topologyId))
+            .isNext(topologyId, sequenceNumber);
+   }
+
+   private boolean checkAllEntries(Map<Integer, Long> sequenceNumbers, int commandTopologyId) {
+      Iterator<Map.Entry<Integer, Long>> iterator = sequenceNumbers.entrySet().iterator();
+      boolean result = true;
+      while (result && iterator.hasNext()) {
+         Map.Entry<Integer, Long> entry = iterator.next();
+         result = checkIfNext(entry.getKey(), commandTopologyId, entry.getValue());
+      }
+      return result;
+   }
+
+   private void checkTopologyId(int topologyId) {
+      if (topologyId != currentCacheTopology.getTopologyId()) {
+         throw OutdatedTopologyException.getCachedInstance();
+      }
+   }
+
+   private class SenderSequencer {
+      @GuardedBy("this")
+      private int topologyId;
+      @GuardedBy("this")
+      private long sequenceNumber = 1;
+
+      private SenderSequencer(int topologyId) {
+         setTopologyId(topologyId);
+      }
+
+      private synchronized long next(int commandTopologyId) {
+         if (topologyId == commandTopologyId) {
+            return sequenceNumber++;
+         } else if (topologyId < commandTopologyId) {
+            //update topology. this command will be the first
+            setTopologyId(commandTopologyId);
+            return sequenceNumber++;
+         } else {
+            //this topology is higher than the command topology id.
+            //another topology was installed. this command will fail with OutdatedTopologyException.
+            throw OutdatedTopologyException.getCachedInstance();
+         }
+      }
+
+      private void setTopologyId(int topologyId) {
+         this.topologyId = topologyId;
+         this.sequenceNumber = 1;
+      }
+   }
+
+   private class ReceiverSequencer {
+      @GuardedBy("this")
+      private int topologyId;
+      @GuardedBy("this")
+      private long nextSequenceNumber = 1;
+
+      private ReceiverSequencer(int topologyId) {
+         setTopologyId(topologyId);
+      }
+
+      private synchronized void deliver(int commandTopologyId, long sequenceNumber) {
+         if (topologyId == commandTopologyId && nextSequenceNumber == sequenceNumber) {
+            nextSequenceNumber++;
+         }
+      }
+
+      private synchronized boolean isNext(int commandTopologyId, long sequenceNumber) {
+         if (topologyId == commandTopologyId) {
+            return nextSequenceNumber == sequenceNumber;
+         } else if (topologyId < commandTopologyId) {
+            //update topology. this command will be the first
+            setTopologyId(commandTopologyId);
+            return nextSequenceNumber == sequenceNumber;
```

Comment:
This branch doesn't seem right. We already check that the command's topology id is the same as the current topology id in `next()` (a bit excessively, I would say), and I assume we also rely on `CheckTopologyAction` to run the command only after the new topology was installed. So I'd rather throw an exception if we got at this point with `topologyId < commandTopologyId`.